File: sys/petscOptionsGetReal.m
function [value, found, errCode, toplevel] = petscOptionsGetReal(pre, name)
% Gets the double precision value for a particular option in the database.
%
%   [value, found, errCode, toplevel] = petscOptionsGetReal(pre, name)
%   obtains a real value in the data base. The flag found is PETSC_TRUE
%   if the attribute was found. The strings must be null-terminated.
%
% SEE ALSO: petscOptionsInsertString, petscOptionsHasName, 
%           PetscOptionsGetInt, PetscOptionsGetString
%
% PETSc C interface:
%   PetscErrorCode  PetscOptionsGetReal(PetscOptions options,const char pre[],const char name[],PetscReal *dvalue,PetscBool  *set)
% http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscOptionsGetReal.html

%#codegen -args {coder.typeof(char(0), [1,inf]), coder.typeof(char(0), [1,inf])}

errCode = int32(-1);

if ~isempty(coder.target)
    toplevel = nargout>3;
    
    if ischar(pre) && ~isempty(pre) && pre(end) && (toplevel || m2c_debug)
        m2c_error('Petsc4m:petscOptionsGetInt:InputError', ...
            'Argument prefix must be a null-terminated string.')
    end
    if ischar(name) && ~isempty(name) && name(end) && (toplevel || m2c_debug)
        m2c_error('Petsc4m:petscOptionsGetInt:InputError', ...
            'Argument name must be a null-terminated character string.')
    end
    value = 0;
    b_flag = coder.opaque('PetscBool');
    
    errCode = coder.ceval('PetscOptionsGetReal', PETSC_NULL_OPTIONS, ...
        coder.rref(pre), coder.rref(name), coder.wref(value), coder.wref(b_flag));
    
    if nargout>1
        found = int32(0); %#ok<NASGU>
        found = coder.ceval('(int)', b_flag);
    end
    
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsGetReal returned error code %d\n', errCode)
    end
end

end

File: sys/codegen/lib/petscOptionsGetReal/petscOptionsGetReal.c
#include "petscOptionsGetReal.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(void);
static void c_m2c_error(int varargin_3);
static void m2c_error(void);
static void b_m2c_error(void)
{
  const char * msgid;
  const char * fmt;
  msgid = "Petsc4m:petscOptionsGetInt:InputError";
  fmt = "Argument name must be a null-terminated character string.";
  M2C_error(msgid, fmt);
}

static void c_m2c_error(int varargin_3)
{
  const char * msgid;
  const char * fmt;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsGetReal returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char * msgid;
  const char * fmt;
  msgid = "Petsc4m:petscOptionsGetInt:InputError";
  fmt = "Argument prefix must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void emxInitArray_char_T(emxArray_char_T **pEmxArray, int numDimensions)
{
  emxInit_char_T(pEmxArray, numDimensions);
}

void petscOptionsGetReal(const emxArray_char_T *pre, const emxArray_char_T *name,
  double *value, int *found, int *errCode, boolean_T *toplevel)
{
  PetscBool b_flag;
  PetscOptions obj;
  *toplevel = true;
  if ((!(pre->size[1] == 0)) && (pre->data[pre->size[1] - 1] != '\x00')) {
    m2c_error();
  }

  if ((!(name->size[1] == 0)) && (name->data[name->size[1] - 1] != '\x00')) {
    b_m2c_error();
  }

  obj = NULL;
  *errCode = PetscOptionsGetReal(obj, &pre->data[0], &name->data[0], value,
    &b_flag);
  *found = (int)(b_flag);
  if (*errCode != 0) {
    c_m2c_error(*errCode);
  }
}

void petscOptionsGetReal_initialize(void)
{
}

void petscOptionsGetReal_terminate(void)
{
}
